airflow/bin/cli.py

```diff
@@ -769,13 +769,58 @@ def worker(args):
         stdout.close()
         stderr.close()
     else:
-        signal.signal(signal.SIGINT, sigint_handler)
-        signal.signal(signal.SIGTERM, sigint_handler)
+        serve_logs_proc = subprocess.Popen(['airflow', 'serve_logs'], env=env)
+        celery_proc = subprocess.Popen(
+            [
+                'celery',
+                'worker',
+                '-A', 'airflow.executors.celery_executor',
+                '-O', 'fair',
+                '-Q', str(args.queues),
+                '-c', str(args.concurrency)
+            ],
+            env=env,
+            # don't inherit stdin, so that Ctrl-C is not handled twice
+            stdin=subprocess.PIPE
+        )
 
-        sp = subprocess.Popen(['airflow', 'serve_logs'], env=env)
 
-        worker.run(**options)
-        sp.kill()
+        def kill_proc(dummy_signum, dummy_frame):
+            serve_logs_proc.terminate()
+
+            # 1. Celery responsd to SIGTERM with warm shutdown quit when all
+            #    tasks finish
+            logging.debug('Sending SIGTERM to Celery')
+            celery_proc.terminate()
+
+            # 2. Send SIGTERM to each child of celery - give up on task.
```

Comment:
We send the celery worker the SIGTERM so it goes into warm shutdown mode (and finishes existing tasks) and spin up a new celery worker to process future tasks.